dali/pipeline/workspace/workspace.h

```diff
@@ -639,150 +619,81 @@ class WorkspaceBase : public ArgumentWorkspace {
+    DALI_ENFORCE_VALID_INDEX(idx, NumOutput());
+    DALI_ENFORCE(outputs_[idx].device == StorageDevice::CPU,
+      make_string("The output ", idx, " is not on the requested device (CPU)."));
+    return outputs_[idx].cpu;
   }
 
   const DataObjectPtr<GPUBackend>& OutputHandle(int idx, const GPUBackend&) const {
-    return GPUOutput(idx);
-  }
-
-  inline const DataObjectPtr<GPUBackend>& GPUInput(int idx) const {
-    auto tensor_meta = FetchAtIndex(input_index_map_, idx);
-    DALI_ENFORCE(tensor_meta.storage_device == StorageDevice::GPU, "Input with given "
-        "index (" + std::to_string(idx) +
-        ") does not have the calling backend type (GPUBackend)");
-    return gpu_inputs_[tensor_meta.index];
-  }
-
-  inline const DataObjectPtr<CPUBackend>& CPUInput(int idx) const {
-    auto tensor_meta = FetchAtIndex(input_index_map_, idx);
-    DALI_ENFORCE(tensor_meta.storage_device == StorageDevice::CPU, "Input with given "
-        "index (" + std::to_string(idx) +
-        ") does not have the calling backend type (CPUBackend)");
-    return cpu_inputs_[tensor_meta.index];
-  }
-
-  inline const DataObjectPtr<GPUBackend>& GPUOutput(int idx) const {
-    auto tensor_meta = FetchAtIndex(output_index_map_, idx);
-    DALI_ENFORCE(tensor_meta.storage_device == StorageDevice::GPU,
-                 make_string("Output with given index (", idx,
-                             ") does not have the calling backend type (GPUBackend)"));
-    return gpu_outputs_[tensor_meta.index];
-  }
-
-  inline const DataObjectPtr<CPUBackend>& CPUOutput(int idx) const {
-    auto tensor_meta = FetchAtIndex(output_index_map_, idx);
-    DALI_ENFORCE(tensor_meta.storage_device == StorageDevice::CPU,
-                 make_string("Output with given index (", idx,
-                             ") does not have the calling backend type (CPUBackend)"));
-    return cpu_outputs_[tensor_meta.index];
+    DALI_ENFORCE_VALID_INDEX(idx, NumOutput());
+    DALI_ENFORCE(outputs_[idx].device == StorageDevice::GPU,
+      make_string("The output ", idx, " is not on the requested device (GPU)."));
+    return outputs_[idx].gpu;
   }
 
   /**
    * Batch sizes for given input indices
    */
   SmallVector<int, 4> batch_sizes_;
 
-  vector<DataObjectPtr<CPUBackend>> cpu_inputs_;
-  vector<DataObjectPtr<CPUBackend>> cpu_outputs_;
-  vector<DataObjectPtr<GPUBackend>> gpu_inputs_;
-  vector<DataObjectPtr<GPUBackend>> gpu_outputs_;
-
-  // Maps from a Tensor position in its typed vector
-  // to its absolute position in the workspaces outputs
-  vector<int> cpu_inputs_index_, gpu_inputs_index_;
-  vector<int> cpu_outputs_index_, gpu_outputs_index_;
-  // Used to map input/output tensor indices (0, 1, ... , num_input-1)
-  // to actual tensor objects. The first element indicates if the
-  // Tensor is stored on cpu, and the second element is the index of
-  // that tensor in the {cpu, gpu}_inputs_ vector.
-  vector<InOutMeta> input_index_map_, output_index_map_;
```

Comment:
The PR is mostly about getting rid of this (and associated functions that manipulate those).